lib/maths/CDataFrameCategoryEncoder.cc

```diff
@@ -190,169 +211,401 @@ class CMinRedundancyMaxRelevancyGreedySearch {
+    } while (traverser.next());
+    return true;
 }
 
-std::size_t CDataFrameCategoryEncoder::numberFeatures() const {
-    return m_FeatureVectorMics.size();
+CDataFrameCategoryEncoder::CIdentityEncoding::CIdentityEncoding(std::size_t inputColumnIndex,
+                                                                double mic)
+    : CEncoding{inputColumnIndex, mic} {
 }
 
-std::size_t CDataFrameCategoryEncoder::encoding(std::size_t index) const {
-    return m_FeatureVectorEncodingMap[index];
+EEncoding CDataFrameCategoryEncoder::CIdentityEncoding::type() const {
+    return E_IdentityEncoding;
 }
 
-std::size_t CDataFrameCategoryEncoder::column(std::size_t index) const {
-    return m_FeatureVectorColumnMap[index];
+double CDataFrameCategoryEncoder::CIdentityEncoding::encode(double value) const {
+    return value;
 }
 
-bool CDataFrameCategoryEncoder::isBinary(std::size_t index) const {
-    std::size_t encoding{this->encoding(index)};
-    std::size_t feature{this->column(index)};
-    return encoding < m_OneHotEncodedCategories[feature].size();
+bool CDataFrameCategoryEncoder::CIdentityEncoding::isBinary() const {
+    return false;
 }
 
-std::size_t CDataFrameCategoryEncoder::numberOneHotEncodedCategories(std::size_t feature) const {
-    return m_OneHotEncodedCategories[feature].size();
+std::uint64_t CDataFrameCategoryEncoder::CIdentityEncoding::checksum() const {
+    return CChecksum::calculate(this->inputColumnIndex(), this->mic());
 }
 
-bool CDataFrameCategoryEncoder::usesOneHotEncoding(std::size_t feature,
-                                                   std::size_t category) const {
-    return std::binary_search(m_OneHotEncodedCategories[feature].begin(),
-                              m_OneHotEncodedCategories[feature].end(), category);
+std::string CDataFrameCategoryEncoder::CIdentityEncoding::typeString() const {
+    return IDENTITY_ENCODING_TAG;
+}
+
+CDataFrameCategoryEncoder::COneHotEncoding::COneHotEncoding(std::size_t inputColumnIndex,
+                                                            double mic,
+                                                            std::size_t hotCategory)
+    : CEncoding{inputColumnIndex, mic}, m_HotCategory{hotCategory} {
+}
+
+EEncoding CDataFrameCategoryEncoder::COneHotEncoding::type() const {
+    return E_OneHot;
+}
+
+double CDataFrameCategoryEncoder::COneHotEncoding::encode(double value) const {
+    return static_cast<std::size_t>(value) == m_HotCategory;
+}
+
+bool CDataFrameCategoryEncoder::COneHotEncoding::isBinary() const {
+    return true;
+}
+
+std::uint64_t CDataFrameCategoryEncoder::COneHotEncoding::checksum() const {
+    std::size_t seed{CChecksum::calculate(this->inputColumnIndex(), this->mic())};
+    return CChecksum::calculate(seed, m_HotCategory);
+}
+
+void CDataFrameCategoryEncoder::COneHotEncoding::acceptPersistInserter(
+    core::CStatePersistInserter& inserter) const {
+    CEncoding::acceptPersistInserter(inserter);
```

Comment:
nit
```suggestion
    this->CEncoding::acceptPersistInserter(inserter);
```

Reply:
I'd be inclined to swap this round though, i.e. have a `acceptPersistInserter` as concrete on `CEncoder` and have a virtual type which persists the rest, i.e. have a private method like:
```
virtual void CEncoding::derivedAcceptPersistInserter(core::CStatePersistInserter& inserter) = 0;
```
The concrete implementation on `CEncoder` can then make sure that the derived type persister is called.

Reply:
done.